Advanced search filters improvements

ginold opened the pull request:
probably fixes #692 
related to #655, #657


Some filters improvements:

+ stickyFilters directive and controller
+ when you first open 'more filters' and scroll down the list - it will not jump and become a 'sticky filter'.
+ only when you scroll down the list without first opening the dropdown - it will become a 'sticky filter'
+ solves the problem where 'activities' sometimes became unusable 
+ solves the problem with reaching the list's bottom 'show results'
+ one filter per line on mobile
+ hide the map on mobile by default, let the user toggle it with a button
+ dropdown menus FINALLY don't overflow to the right under he map - they are repositioned (as it was done for image uploader).
+ dropdown menus are not position: fixed - they move with the list as you scroll (more natural)

![screenshot from 2016-10-26 11-12-49](https://cloud.githubusercontent.com/assets/7814311/19720130/280cf29c-9b6d-11e6-9923-a040ff1c4a5f.png)

This pr only works for routes, but before I add the changes for other documents, I would like to know it what i'm doing is ok.

asaunier commented:
Please rebase your PR to upgrade your dependencies ("moment"). It seems the reason why snyk complains.

asaunier commented:
Is the PR ready for reviewing? Still need a rebase anyway :P

ginold commented:
rebased. You can start reviewing while i'm working on a few things in this pr.

asaunier commented:
I see you have made changes to the route filters. Don't you need to apply such changes to other document types?

ginold commented:
maybe yes. I just moved 2 filters above because the dropdown overflows the bottom of the screen. As i said in the description - i did changes only for the routes for now. I'm waiting for your confirmation to know if what i'm doing is ok.

asaunier commented:
Don't your changes break the other doc types anyway?

As far as I can see, your PR looks OK to me. I will test your branch in my instance.

ginold commented:
yes they do, I mean i created a directive that I haven't put in the other templates yet. But if you told me this is completely wrong etc, I would have done some additional work for nothing :P

asaunier commented:
My concern is about the map filtering. What happens when the map is hidden? I have to test it live.

ginold commented:
Don't worry - Your doubts would be adequate if the map had `display: none;` - it would in fact not work and the bounding box would never be generated. But i use `opacity: 0;` which means that the map is still virtually there, but transparent, invisible.

asaunier commented:
@ginold I have tested your branch in my instance and it works really great, including in mobile mode. You rulezzzzzzz :)

Since the map is initially "transparent" on a mobile screen, what about making it use the whole screen when finally displayed? As for now it uses only half or 2 thirds of the screen height. Anyway it's hard to see both the cards and the map at the same time.

![screencapture-localhost-6555-routes-1477569719472](https://cloud.githubusercontent.com/assets/1192331/19766426/0798d03a-9c4e-11e6-972b-f153107d273c.png)

ginold commented:
updated pr with the so long-awaited feature: 

+ dropdown menus FINALLY don't overflow to the right under he map - they are repositioned (as it was done for image uploader).
+ dropdown menus are not position: fixed - they move with the list as you scroll (more natural)

there are some bugs, but they don't appear every time. A deep testing is welcome!

asaunier commented:
Perhaps I'm getting old but I have found nothing special to complain about in the code. Perhaps you're etting better :)

Anyway I think it's an excellent work!

Concerning the behaviour, I have found a few minor problems (perhaps already there before your changes), but nothing critical. If you cannot find easy and fast fixes, I think it's OK to merge the PR anyway and to list the problems in a new issue:
* in mobile screens the map could be bigger (as discussed above) - not critical anyway
* in mobile screens when there is less than < 30 results, the pagination bar is not shown, if the map is open one can see the cards in the background below the map - not critical anyway
* in mobile screens sometimes I have observed that that "see results" "clear filters" buttons were shifted to the left and not fully visible. But perhaps I had resized the screen in the meantime? I cannot reproduce the problem. See the screenshot below:
![screencapture-localhost-6555-waypoints-1477588526266](https://cloud.githubusercontent.com/assets/1192331/19778603/d1def458-9c7d-11e6-8fec-588904dd32b1.png)
* when scrolling down the list, the "sticky" buttons appear at the top of the list. If I click on the "all filters" button and if there are too many filters available, it's not possible to scroll down the filters and click on the "show results" or "clear filters" buttons. I think this problem is not related to this PR (see https://github.com/c2corg/v6_ui/issues/692)
* when searching doctypes with no map filter (eg. articles), in mobile screens since the "map" button in the bottom right corner is not displayed, there is a white gap at the right hand of the pagination bar (OK, really minor :P)

ginold commented:
Updated! Hopefully I fixed all the bugs you mentioned. You can retest.

asaunier commented:
* [ ] a problem with double-slider filters in the primary filters section:
 * move an handle of a slider to some position
 * scroll the list until the "sticky" buttons appear at the top of the list
 * click on "see all filters"
 * the slider filter that is then shown is not in the expected status (handle not at the right position)
 * if clicking on "view results", the original slider is shown in the expected status
 * no problem with other filter types (text input or checkbox panel)
 * no problem if clicking on "see all filters" when the primary section is shown

asaunier commented on c2corg_ui/static/partials/advancedsearch.html:
```diff
@@ -4,3 +4,5 @@
               ng-mouseleave="searchCtrl.onMouseLeave(doc.document_id)"></app-card>
   </div>
 </div>
+<p translate class="text-warning text-center no-results" ng-class="{'show': searchCtrl.noResults}">No results</p>
```
A dream come true! Thanks :)

asaunier commented:
I have only tested with Chromium. Looks to work fine except the comment above.
Is it OK with other browsers?
For instance this ``.calc()`` things in the styles? I guess so: https://developer.mozilla.org/fr/docs/Web/CSS/calc#Compatibilité_des_navigateurs

Thanks for enlarging the map. Not a big deal but its size is still a bit weird, especially at the top:
![screencapture-localhost-6555-routes-1477652483834](https://cloud.githubusercontent.com/assets/1192331/19805013/2b604ce0-9d12-11e6-8c01-0877a05a656d.png)

ginold commented:
>  For instance this .calc() things in the styles?

I hope so. This is just a function that returns 

```
.calc(@prop; @val) {
  @{prop}: calc(~'@{val}');
  @{prop}: -moz-calc(~'@{val}');
  @{prop}: -webkit-calc(~'@{val}');
  @{prop}: -o-calc(~'@{val}');
}
```

So it is adapted to all browsers.

ginold commented:
>  a problem with double-slider filters in the primary filters section

this is annoying, true. I think the two sliders found in .simple-search and all filters don't sync. There is probably some work to do with the slider-directive, when the slide occurs it should sync across the same type of slider. Maybe for another PR?

asaunier commented:
> Maybe for another PR? This is not critical if one slider doesn't sync (it's ok for both on page load thanks to the data from url).

OK for me. Please merge and add this problem to issue https://github.com/c2corg/v6_ui/issues/655
Thanks!
